### PR TITLE
feat(route-recognizer): Support generating a route by config as well as name

### PR DESCRIPTION
**Pull request by davismj**
This change allows the router to generate routes for all configured routes, not only those with name, which allows the router.generate method to be used reliably by the router internals.

### CODE REVIEW

File: src/route-recognizer.js [hunk 1]
```diff
@@ -79,13 +80,13 @@ export class RouteRecognizer {
...
 
+    this.routes.set(route.handler, { segments, handlers });
     if (routeName) {
       let routeNames = Array.isArray(routeName) ? routeName : [routeName];
       for (let i = 0; i < routeNames.length; i++) {
-        this.names[routeNames[i]] = {
-          segments: segments,
-          handlers: handlers
-        };
+        if (!(routeNames[i] in this.names)) {
```
Comment (davismj) [inline src/route-recognizer.js]: I forgot to document this change. If two routes had the same name, e.g. `{ route: ['a','b'], name: 'foo' }`, which is processed as two routes, the latter route would override the former in the name index. This means if you would generate a route for 'foo' above, you would get '#/b' as a result. 

This was a regression. I'm not sure where the regression was introduced, however.

**Comment by davismj** [comment]
### Support me on Patreon

I love working to make Aurelia the greatest front-end framework in the world. If you like my work and want to support me in this mission, please consider backing me on Patreon so I can afford to make more contributions like this one. Thank you!

<a href="http://patreon.com/davismj"></a>

**Comment by bigopon** [comment]
lgtm. Though I'm not sure about the name of `RouteGenerator`

cc @EisenbergEffect

**Comment by EisenbergEffect** [comment]
Looks good.

**Pull request merged**